src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -454,10 +484,45 @@ protected virtual void CompleteTransaction(bool isCommitted)
 				// do an early exit here in such case.
 				if (!IsInActiveTransaction)
 					return;
+				Lock();
+				// In case of a rollback due to a timeout, we may have the session disposal running concurrently
+				// to the transaction completion in a way our current locking mechanism cannot fully protect: the
+				// session disposal "BeginProcess" can go through the Wait before it is locked but flag the
+				// session as procesisng after the transaction completion has read it as not processing. To dodge
+				// that case, we consider the session as still processing initially regardless of its actual
+				// status in case of rollback.
+				var isSessionProcessing = !isCommitted || _session.IsProcessing();
 				try
 				{
 					// Allow transaction completed actions to run while others stay blocked.
 					_bypassLock.Value = true;
```

Comment:
Maybe the proper fix would be to remove bypass lock?

Reply:
What happens if before Lock() we put Wait()?

Reply:
Systematic deadlock then lock timeout with thrown exception, for both questions.

If the transaction is no more considered active, `Wait` activates the lock, which will block until deactivated, whatever the thread. Wait then attempt to go through the lock unless its thread has bypass lock set. Bypass lock allows the transaction completion thread to not get blocked by the lock.